[CALCITE-5583] JDBC adapter does not generate 'SELECT *' when duplicate field names

xiedeyantu opened the pull request:
Jira: [CALCITE-5583](https://issues.apache.org/jira/browse/CALCITE-5583)

xuzifu666 commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -8774,6 +8941,67 @@ private void checkLiteral2(String expression, String expected) {
+            ImmutableList.of("c0", "c00"),
+            true)
+        .build();
+    // LogicalProject(c0=[$0], c00=[$1])
```
Comments here should be more details or remove it?

julianhyde commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -4531,8 +4575,8 @@ private SqlDialect nonOrdinalDialect() {
-    final String expected = "SELECT reserve_employee.hire_date, "
-        + "CAST(reserve_employee.hire_date AS VARCHAR(10))\n"
+    final String expected = "SELECT reserve_employee.hire_date,"
+        + " CAST(reserve_employee.hire_date AS VARCHAR(10))\n"
```
cosmetic change. you should revert

xiedeyantu commented:
I have modified the previous logic, which mainly includes two aspects:
1. Added a supportGenerateSelectStar method in SqlDialect to control the behavior of each dialect. Currently, only PostgreSQL has special handling. This way, other test cases in RelToSqlConverterTest don’t need to be modified.
2. Removed the previously added isDistinct method and moved the case-sensitivity check earlier, while still using the original isDistinct method for the actual judgment.
@suibianwanwank @julianhyde Please help to take a look again. Thanks!

NobiGo commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -8772,6 +8772,64 @@ private void checkLiteral2(String expression, String expected) {
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-5583">[CALCITE-5583]
+   * Rel2Sql will get an error when select * and join is present</a>. */
+  @Test void testSelectStarWithJoinNatural() {
```
This is an invalid test case. Regardless of whether this PR exists or not, it can still be executed successfully.

NobiGo replied:
I suggest first checking if any other test cases have already covered this aspect.

NobiGo commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -8772,6 +8772,64 @@ private void checkLiteral2(String expression, String expected) {
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-5583">[CALCITE-5583]
+   * Rel2Sql will get an error when select * and join is present</a>. */
+  @Test void testSelectStarWithJoinOn() {
+    final String sql = "SELECT * FROM \"DEPT\" t1 "
```
Can this SQL be executed directly in PG?

xiedeyantu replied:
I verified that a similar example can run on a pgSQL web page. [test for pgsql](https://onecompiler.com/postgresql/43n9x23hm)

NobiGo replied:
Get it. Thanks.

NobiGo commented on core/src/main/java/org/apache/calcite/sql/SqlDialect.java:
```diff
@@ -797,6 +797,11 @@ public boolean supportBooleanCaseWhen() {
   }
 
+  /** Returns whether this dialect supports generate 'SELECT *' when duplicate field names. */
+  public boolean supportGenerateSelectStar() {
```
The comment for this method is somewhat ambiguous. From an implementation perspective, it is not solely controlled by this method.

xiedeyantu replied:
I agree with your comment. Indeed, I didn't come up with a good method name and code comments here. I will think about it again.

NobiGo replied:
I suggest passing in the RelNode as a parameter, and then different Dialects can make more customized decisions based on their specific circumstances regarding whether to generate star.

xiedeyantu replied:
Great suggestion, this makes the code logic clearer. I will revise according to this suggestion later.

NobiGo reviewed (approved):
LGMT
